src/Main.elm

```diff
@@ -25,40 +143,939 @@ init =
+                        { formData
+                            | tags = updatedTags
+                            , inputFields = resetTagsInputs
+                        }
+                in
+                ( { model | errors = Set.empty, formData = updatedFormData }, Cmd.none )
+
+        EditTag tagName ->
+            let
+                { formData } =
+                    model
+
+                { tags, inputFields } =
+                    formData
+            in
+            case Dict.get tagName tags of
+                Just tagValue ->
+                    let
+                        updateInputFields =
+                            inputFields
+                                |> Dict.insert "newTagName" tagName
+                                |> Dict.insert "newTagValue" tagValue
+
+                        updatedFormData =
+                            { formData | inputFields = updateInputFields }
+                    in
+                    ( { model | formData = updatedFormData }, Cmd.none )
+
+                Nothing ->
+                    ( model, Cmd.none )
+
+        DeleteTag tagName ->
+            let
+                { formData } =
+                    model
+
+                { tags } =
+                    formData
+
+                updatedTags =
+                    tags |> Dict.remove tagName
+
+                updatedFormData =
+                    { formData | tags = updatedTags }
+            in
+            ( { model | formData = updatedFormData }, Cmd.none )
+
+
+formValidate :
+    { inputFields : Dict String String
+    , prefMethod : PreferredContactMethod
+    , isDetailsInherited : Bool
+    , isSettingsInherited : Bool
+    }
+    -> Set String
+formValidate { inputFields, prefMethod, isDetailsInherited, isSettingsInherited } =
+    let
+        emailValue =
+            inputFields |> Dict.get "email" |> Maybe.withDefault ""
+
+        phoneValue =
+            inputFields |> Dict.get "phone" |> Maybe.withDefault ""
+
+        settingsList =
+            [ "preparation", "closed", "canceled", "timedOut", "rejected", "error" ]
+
+        updateErrors =
+            let
+                detailsErrors =
+                    -- We'll validate details only if not inherited
+                    if isDetailsInherited then
+                        Set.empty
+
+                    else
+                        let
+                            emailError =
+                                let
+                                    errorMsg =
+                                        "Details Tab: Invalid email"
+                                in
+                                case Util.parseEmail (inputFields |> Dict.get "email" |> Maybe.withDefault "") of
+                                    Ok _ ->
+                                        Set.remove errorMsg Set.empty
+
+                                    Err _ ->
+                                        Set.insert errorMsg Set.empty
```

Comment:
NIT: This seems like a really deeply nested let..in blocks, which is hard to keep track of what context are you in. I would probably prefer the to be chunked and refactored away

Reply:
Yeah, it was a very annoying part to validate all the fields in one place. Took that decision so I don't do validation onChange per a field, since in that case (I think) I would need to write all OnChange updates for each fields. So took "less evil" path.